draft-ietf-mls-protocol.md

```diff
@@ -1008,14 +1112,178 @@ struct {
+    handshake(1),
+    application(2),
+    (255)
+} ContentType;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    uint32 sender;
+    ContentType content_type;
+
+    select (MLSPlaintext.type) {
+        case handshake:
+            GroupOperation operation;
+
+        case application:
+            opaque application_data<0..2^32-1>;
+    }
+
+    opaque signature<0..2^16-1>;
+} MLSPlaintext;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    ContentType content_type;
+    opaque masked_sender_data[8];
+    opaque ciphertext<0..2^32-1>;
+} MLSCiphertext;
+~~~~~
+
+The remainder of this section describe how to compute the signature of
+an MLSPlaintext object and how to convert it to an MLSCiphertext object.
+The overall process is as follows:
+
+* Compute the metadata by concatenating the following information:
+  * Group ID
+  * Epoch
+  * Sender index
+  * Key generation
+
+* Sign the protected content and metadata
+
+* Encrypt the content using a content encryption key identified by
+  the metadata
+
+* Mask the sender_data used to identify the content encryption key
+  using a per-epoch key
+
+The group identifier and epoch fields are copied from the MLSPlaintext object directly.
+The content encryption process populates the ciphertext field of the
+MLSCiphertext object.  The metadata encryption step populates the
+masked_sender_data field.
+
+Decryption follows the same step in reverse: Unmasking the
+metadata, then decrypting the content, then verifying the content
+signature.
+
+
+## Content Signing and Encryption
+
+The signature field in an MLSPlaintext object is computed using the
+signing private key corresponding to the credential at the leaf in
+the tree indicated by the sender field.  The signature covers the
+MLSPlaintext object, with the signature field truncated.
+
+The ciphertext field of the MLSCiphertext object is produced by
+supplying the inputs described below to the AEAD function specified
+by the ciphersuite in use.  The plaintext input contains content and
+signature of the MLSPlaintext, plus optional padding.  These values
+are encoded in the following form:
+
+~~~~~
+struct {
+    opaque content[MLSPlaintext.length];
+    uint8 signature[MLSInnerPlaintext.sig_len];
+    uint16 sig_len;
+    uint8  marker = 1;
+    uint8  zero\_padding[length\_of\_padding];
+} MLSCiphertextContent;
+~~~~~
+
+The key and nonce used for this encryption depend on the content
+type of the message.  The sender chooses an handshake_key for an
+handshake message or an ununsed generation from its (per-sender)
+application key chain for the current epoch, according to the type
+of message being encrypted.
+
+The Additional Authenticated Data (AAD) input to the encryption
+contains an object of the following form, with the values used to
+identify the key and nonce:
+
+~~~~~
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    ContentType content_type;
+    opaque masked_sender_data[8];
```

Comment:
This is impossible.  The masking is based on the ciphertext, so you can't use the masked values as an input to the encryption.  Just use the plain values.